lib/memoryprofiler/derive/src/lib.rs

```diff
@@ -0,0 +1,62 @@
+extern crate proc_macro;
+use quote::quote;
+use syn::*;
+
+#[proc_macro_derive(MemoryUsage)]
+pub fn derive_memory_usage(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let input: DeriveInput = syn::parse(input).unwrap();
+    match input.data {
+        Data::Struct(ref struct_data) => {
+            derive_memory_usage_struct(&input.ident, struct_data, &input.generics)
+        }
+        _ => unreachable!("not yet implemented"),
+        /*
+        Data::Enum(ref enum_data) => {
+            derive_memory_usage_struct(enum_data)
+        },
+        Data::Union(ref union_data) => {
+            derive_memory_usage_union(union_data)
+        },
+        */
+    }
+}
+
+fn derive_memory_usage_struct(
+    struct_name: &Ident,
+    data: &DataStruct,
+    generics: &Generics,
+) -> proc_macro::TokenStream {
+    let lifetimes_and_generics = &generics.params;
+    let where_clause = &generics.where_clause;
+    let sum = match &data.fields {
+        Fields::Named(ref fields) => fields
+            .named
+            .iter()
+            .map(|field| {
+                let id = field.ident.as_ref().unwrap();
+                quote! { MemoryUsage::size_of(&self.#id) }
```

Comment:
You can use `quote_spanned` and get the span from the field so that way if there's an error message (for example one field does not implement `MemoryUsage`) the compiler will point to the line containing that field and it'll show that field in the error message

Reply:
Done.

Do you know any way to do this for `quote! ( x + y )` where I have one span for `x` and another for `y`?